probes/memorysafe/impl.go

```diff
@@ -0,0 +1,229 @@
+)
+
+//go:embed *.yml
+var fs embed.FS
+
+const (
+	Probe = "memorysafe"
+)
+
+type languageMemoryCheckConfig struct {
+	Desc string
+
+	funcPointers []func(client *checker.CheckRequest) ([]finding.Finding, error)
+}
+
+var languageMemorySafeSpecs = map[clients.LanguageName]languageMemoryCheckConfig{
+	clients.Go: {
+		funcPointers: []func(client *checker.CheckRequest) ([]finding.Finding, error){
+			checkGoUnsafePackage,
+		},
+		Desc: "Check if Go code uses the unsafe package",
+	},
+
+	clients.CSharp: {
+		funcPointers: []func(client *checker.CheckRequest) ([]finding.Finding, error){
+			checkDotnetAllowUnsafeBlocks,
+		},
+		Desc: "Check if C# code uses unsafe blocks",
+	},
+}
+
+func init() {
+	probes.MustRegisterIndependent(Probe, Run)
+}
+
+func Run(raw *checker.CheckRequest) (found []finding.Finding, probeName string, err error) {
+	prominentLangs := getRepositoryLanguageChecks(raw)
+	findings := []finding.Finding{}
+
+	for _, lang := range prominentLangs {
+		for _, langFunc := range lang.funcPointers {
+			if langFunc == nil {
+				raw.Dlogger.Warn(&checker.LogMessage{
+					Text: fmt.Sprintf("no function pointer found for language %s", lang.Desc),
+				})
+			}
+			langFindings, err := langFunc(raw)
+			if err != nil {
+				return nil, Probe, fmt.Errorf("error while running function for language %s: %w", lang.Desc, err)
+			}
+			findings = append(findings, langFindings...)
+		}
+	}
+	return findings, Probe, nil
+}
+
+func getRepositoryLanguageChecks(raw *checker.CheckRequest) []languageMemoryCheckConfig {
+	langs, err := raw.RepoClient.ListProgrammingLanguages()
+	if err != nil {
+		raw.Dlogger.Warn(&checker.LogMessage{
+			Text: fmt.Sprintf("RepoClient retured error for ListProgrammingLanguages: %v", err),
+		})
+		return nil
+	}
+	if len(langs) == 0 {
+		return []languageMemoryCheckConfig{}
+	}
+	if len(langs) == 1 && langs[0].Name == clients.All {
+		return getAllLanguages()
+	}
+	ret := []languageMemoryCheckConfig{}
+	for _, language := range langs {
+		if lang, ok := languageMemorySafeSpecs[clients.LanguageName(strings.ToLower(string(language.Name)))]; ok {
+			ret = append(ret, lang)
+		}
+	}
+	return ret
+}
+
+func getAllLanguages() []languageMemoryCheckConfig {
+	allLanguages := make([]languageMemoryCheckConfig, 0, len(languageMemorySafeSpecs))
+	for l := range languageMemorySafeSpecs {
+		allLanguages = append(allLanguages, languageMemorySafeSpecs[l])
+	}
+	return allLanguages
+}
+
+// Golang
+
+func checkGoUnsafePackage(client *checker.CheckRequest) ([]finding.Finding, error) {
+	findings := []finding.Finding{}
+
+	if err := fileparser.OnMatchingFileContentDo(client.RepoClient, fileparser.PathMatcher{
+		Pattern:       "*.go",
+		CaseSensitive: false,
+	}, goCodeUsesUnsafePackage, &findings, client.Dlogger); err != nil {
+		return nil, err
+	}
+	if len(findings) == 0 {
+		found, err := finding.NewWith(fs, Probe,
+			"Golang code does not use the unsafe package", nil, finding.OutcomeTrue)
```

Comment:
we could also move this (and the c# version) to one combined top level check in `Run` before the return,depending on the semantics we want the probe to have.

Reply:
(y)